bpo-38216, bpo-36274: Allow subclasses to override validation and encoding behavior

jaraco opened the pull request:
This patch introduces a new semi-private hook for subclasses to override the validation and encoding behavior in `http.client.HTTPConnection.putrequest`, allowing select clients to customize the behavior and implement invalid requests as was possible prior to the patch in [bpo-30458](https://bugs.python.org/issue30458) (for control characters) and prior to Python 3.0 (for encoding).

<!-- issue-number: [bpo-38216](https://bugs.python.org/issue38216) -->
https://bugs.python.org/issue38216
<!-- /issue-number -->

jaraco committed:
bpo-38216: Allow bypassing input validation

jaraco committed:
bpo-36274: Also allow the URL encoding to be overridden.

jaraco committed:
bpo-38216, bpo-36274: Add tests demonstrating a hook for overriding validation, test demonstrating override encoding, and a test to capture expectation of the interface for the URL.

jaraco committed:
Merge branch '38216-putrequest-hooks-tests' into feature/38216-putrequest-hooks

jaraco committed:
Call with skip_host to avoid tripping on the host checking in the URL.

webknjaz commented on Lib/test/test_httplib.py:
This little hanging thing looks weird. Maybe do smth like
```python
pattern = client._contains_disallowed_url_pchar_re
if match := pattern.search(url):
```

zooba replied:
Or just ignore the line length _guideline_ for this one case.

gpshead replied:
it's fine as is.  using `m` as the name instead of `match` would've been my suggestion. ;)

gpshead commented on Lib/test/test_httplib.py:
Rather than requiring people to override the entire def _prepare_path method just to change the encoding it allows (which, as this example shows, means copy and pasting the first part of the implementation), how about making a `._prepare_path_encoding` attribute to the class that could be overridden.  it'd default to `'ascii'` but you would set it to `'utf-8'` in this subclass.  instead of overriding the method.

jaraco replied:
I considered this, but the proposed change that inspired this change wished to change the invocation of `encode('ascii')` to `encode('ascii', errors='surrogateescape')`, so that seemed like enough of a signal that one might wish to override the encoding process entirely.

jaraco replied:
Does it still make sense to use `self._prepare_path_encoding` so it's easier to override if that's what someone wants?

gpshead replied:
i suppose _prepare_path_encoding could be its own method in that case.  I wanted to avoid people needing to duplicate the control characters check for this situation.

gpshead commented on Lib/http/client.py:
We can get rid of this comment now.  it was referring to the `.encode('ascii')` that is no longer below.

gpshead reviewed (changes requested):
also, please use blurb to add a NEWS entry for this change.

gpshead committed:
Remove obsolete comment.

gpshead committed:
Make _prepare_path_encoding its own attr.

This makes overriding just that simpler.

Also, don't use the := operator to make backporting easier.

gpshead committed:
Add a news entry.

gpshead committed:
_prepare_path_encoding -> _encode_prepared_path()
